File: ui/src/services/FeedbackService.ts
import axios, {AxiosError} from "axios";

export interface FeedbackCardData {
    feedbackId: number,
    firstName: string,
    lastName: string,
    tutorId: number | null,
    studentId: number | null,
    rate: number,
    feedback: string,
    languageDescription: string,
    userAvatarUrl: string
}

export interface FeedbackAboutStudentRequestDto {
    conversationId: number
    tutorRate: number
    tutorFeedback: string
}

export interface FeedbackAboutTutorRequestDto {
    conversationId: number
    studentRate: number
    studentFeedback: string
}

class FeedbackService {
    findFeedbacksAboutTutor = async (id: number): Promise<FeedbackCardData[]> => {
        try {
            const response = await axios.get('http://localhost:8080/api/v1/feedbacks/tutor/'
                + id, {withCredentials: true});
            return response.data;
        } catch (e: unknown) {
            throw e as AxiosError;
        }
    }

    findFeedbacksAboutStudent = async (id: number): Promise<FeedbackCardData[]> => {
        try {
            const response = await axios.get('http://localhost:8080/api/v1/feedbacks/student/'
                + id, {withCredentials: true});
            return response.data;
        } catch (e: unknown) {
            throw e as AxiosError;
        }
    }

    saveFeedbackAboutStudent = async (dto: FeedbackAboutStudentRequestDto) => {
        try {
            const response = await axios.put('http://localhost:8080/api/v1/feedbacks/student/',
                dto, {withCredentials: true});
            return response.status == 201;
        } catch (e: unknown) {
            throw e as AxiosError;
        }
    }

    saveFeedbackAboutTutor = async (dto: FeedbackAboutTutorRequestDto) => {
        try {
            const response = await axios.put('http://localhost:8080/api/v1/feedbacks/tutor/',
                dto, {withCredentials: true});
            return response.status == 201;
        } catch (e: unknown) {
            throw e as AxiosError;
        }
    }
}

export default new FeedbackService();

File: api-gateway/src/main/java/com/example/apigateway/services/FeedbackService.java
package com.example.apigateway.services;

import com.example.apigateway.client.feedback.FeedbackClient;
import com.example.apigateway.dto.FeedbackAboutStudentRequestDto;
import com.example.apigateway.dto.FeedbackAboutStudentResponseDto;
import com.example.apigateway.dto.FeedbackAboutTutorRequestDto;
import com.example.apigateway.dto.FeedbackAboutTutorResponseDto;
import com.example.apigateway.model.Feedback;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class FeedbackService {

    private final FeedbackClient feedbackClient;

    public Feedback saveFeedbackAboutStudent(FeedbackAboutStudentRequestDto dto) {
        return feedbackClient.saveFeedbackAboutStudent(dto);
    }

    public Feedback saveFeedbackAboutTutor(FeedbackAboutTutorRequestDto dto) {
        return feedbackClient.saveFeedbackAboutTutor(dto);
    }

    public List<FeedbackAboutTutorResponseDto> findFeedbacksAboutTutor(Long tutorId) {
        return feedbackClient.findFeedbacksAboutTutor(tutorId);
    }

    public List<FeedbackAboutStudentResponseDto> findFeedbacksAboutStudent(Long studentId) {
        return feedbackClient.findFeedbacksAboutStudent(studentId);
    }

}
